### PR TITLE
Fixes #712: Mock (some of the) communication with GitHub API

### LINKED ISSUE #712: Mocking the GitHub communications for functional tests
When we are testing the UI with intern, we want to make sure that the UI reacts accordingly to what we want of the user experience. Right now, the tests are dependent of Github communication but it's not necessary most of the time for the suer experience. Communications issues in between webcompat and github have to be dealt with elsewhere. 

1. Mock all response from Github to WebCompat in intern (when possible)

There is two features which are currently directly talking to Github. 
* search everywhere (in progress)
* load search results from URI bar

decided in Paris september 2015 at webcompat meeting  with @miketaylr

### PR DESCRIPTION
**Pull request by miketaylr**
r? @karlcow 

Here's the general strategy:

We have a `@mockable_repsponse` decorator to add to an API endpoint. 

IF we're in test mode (`python run.py -t`), instead of making the request to GitHub, we serve a file from `test/fixtures/` instead.

Limitations: 
* Creating fixtures files is an annoying manual process. I plan on writing a script to generate/record these in the future. #777.
* We're not yet mocking login/logout w/ GitHub. #776.
* We're only mocking READs from the API and not WRITEs. #778.
* I need to write docs on adding new fixture files until we have a script to generate them. #775.

I needed to remove a few tests from our label tests, but will open an issue to figure that out. Solving the WRITE mocking will help, I think.

### CODE REVIEW

**Comment by miketaylr** [comment]
One failure on the travis push:

```
FAIL: main - Search (non-auth) - Results are loaded from the query params (306ms)
AssertionError: Loading image is visible: expected 'wc-Loader js-loader' to include 'is-active'
```

I was having trouble with this locally as well. The mock response is too fast (like 3 - 5ms), I think it goes away before Intern can find the loader image. This might mean we need to re-think how to test this.

File: webcompat/helpers.py [hunk 1]
```diff
@@ -12,7 +12,11 @@
 import urlparse
 
 from babel.dates import format_timedelta
+from flask import g
+from flask import request
 from flask import session
+from flask import send_from_directory
```
Comment (miketaylr) [inline webcompat/helpers.py]: Ooops, thought I removed this. Unused.

**Comment by karlcow** [comment]
Ah interesting! You decided to mock at the python level. I thought that you would mock at the intern level of the functional test suite, which are the one mostly generating requests to the github API through webcompat. 

https://www.sitepen.com/blog/2014/07/14/mocking-data-with-intern/

Ok let me look at this one. At first sight it looks good. Let me try to import your branch.

``` bash
git checkout master
git fetch upstream;git merge upstream/master
# Importing your branch in a local branch
git checkout -b 712/1 upstream/issues/712/1
# Branch 712/1 set up to track remote branch issues/712/1 from upstream.
# Switched to a new branch '712/1'
python run.py
# * Running on http://localhost:5000/ (Press CTRL+C to quit)
# * Restarting with stat
# OK this part is working
```

In a separate shell window

```bash
java -jar /Users/karl/bin/selenium-server-standalone-2.46.0.jar
# 11:22:19.468 INFO - Launching a standalone Selenium Server
# 11:22:19.545 INFO - Java: Oracle Corporation 24.51-b03
# 11:22:19.545 INFO - OS: Mac OS X 10.11 x86_64
# 11:22:19.578 INFO - v2.46.0, with Core v2.46.0. Built from revision 87c69e2
# 11:22:19.706 INFO - Driver provider org.openqa.selenium.ie.InternetExplorerDriver registration is skipped:
# registration capabilities Capabilities [{platform=WINDOWS, ensureCleanSession=true, browserName=internet explorer, version=}] does not match the current platform MAC
# 11:22:19.707 INFO - Driver class not found: com.opera.core.systems.OperaDriver
# 11:22:19.707 INFO - Driver provider com.opera.core.systems.OperaDriver is not registered
# 11:22:20.134 INFO - RemoteWebDriver instances should connect to: http://127.0.0.1:4444/wd/hub
# 11:22:20.134 INFO - Selenium Server is up and running
```

In yet another shell window:

```bash
workon webcompatcom
# (webcompatcom)11:24:13 ~/code/webcompat.com
node_modules/.bin/intern-runner config=tests/intern
```

I wonder if I'm missing a couple of things. Probably CSS/JS. Some tests are failing. 

capture d ecran 2015-10-21 a 11 29 38

Interesting if I type

```
→ grunt

module.js:340
    throw err;
          ^
Error: Cannot find module 'grunt-legacy-log'
    at Function.Module._resolveFilename (module.js:338:15)
    at Function.Module._load (module.js:280:25)
    at Module.require (module.js:364:17)
    at require (module.js:380:17)
    at Object.<anonymous> (/Users/karl/code/webcompat.com/node_modules/grunt/lib/grunt.js:30:11)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Module.require (module.js:364:17)
```

**Comment by karlcow** [comment]
Another source for doing it on the intern side. I do not know if it's easier.
http://davidwalsh.name/nock
https://github.com/pgte/nock

**Comment by miketaylr** [comment]
The way I have in this branch works, but I think it's kind of weird (and intrusive) to put in the mocking decorators right in the endpoint routes.

> I wonder if I'm missing a couple of things. 

Ah yeah, start the server in test mode: `python run.py -t`. There should only be 1 failure related to a loader image, which I need to fix.

I did look at Nock (and about 8 similar things -- but all my notes are local, oops). It seems like similar things work really well if your app is written in Node, or at least geared towards unit testing (rather than in-browser functional testing). It might be possible to glue this and Intern together though? I'll look around.

Another option I thought about was running the site behind a proxy that intern runs func tests on, and intercepts requests to port 5000 that start with `/api/` returning mock data, passing everything else through transparently. But then I chased after the current approach.

**Comment by miketaylr** [comment]
- [ ] remove unused import
- [ ] fix failing test
- [ ] consider doings something else entirely :p

**Comment by miketaylr** [comment]
(Hmm, I'm not sure about that grunt error though. O_o)

**Comment by karlcow** [comment]
> (Hmm, I'm not sure about that grunt error though. O_o)

yes. Frustrating because the UI of the different branches I'm trying is not working so far. I checked the permissions in node, some were given to root and fixed them with

```bash
sudo chown -R karl:staff node_modules
```

Tried `make update` but it failed too.
```
npm ERR! Additional logging details can be found in:
npm ERR!     /Users/karl/code/webcompat.com/npm-debug.log
npm ERR! not ok code 0
make: *** [update] Error 3
```

Ah maybe more locking in 
```
4734 error Error: EACCES, mkdir '/Users/karl/.npm/lodash/2.4.2'
```

Let's see
```bash
sudo chown -R karl:staff /Users/karl/.npm
```
Gotcha!

`make update` worked.
Let's try `grunt`.

yeeha!
`Done, without errors.`

Maybe I can debug @magsout issue now https://github.com/webcompat/webcompat.com/commit/432ba9ee22ef0a49198cc97507262c0510973a78#commitcomment-13916997

**Comment by miketaylr** [comment]
Heh, computers are the worst.

**Comment by miketaylr** [comment]
Oops, I forgot to run the server in test mode before pushing my "re-write" of that failing test. :/

**Comment by miketaylr** [comment]
Needed to rebase against master to fix a conflict, so incoming destructive `git push -f`.

**Comment by miketaylr** [comment]
OK, tests should be passing now. I filed https://github.com/webcompat/webcompat.com/issues/791 to come back and make sure we add/re-write the test that I ripped out to get this to pass.

@karlcow, what do you think. We merge now and keep improving mocked tests? 

Or not worth the effort and re-write everything w/ a proxy? That's fine, but I probably won't revisit this work for a few months if so--have to move on to other Moz goals stuff and screenshot uploading.

**Comment by karlcow** [comment]
@miketaylr in the back of my head I have the feeling there's a way that would make it easier, but I haven't found it too. So :)

To move forward and to breathe a bit, let's merge this one and we can still improve in the future. It's a step in the direction we desire. :)

:100:

**Comment by miketaylr** [comment]
Cool, thanks. I might play with the proxy approach on some stray Saturday -- I think that will be simpler.

**Pull request merged**